mungegithub/mungers/publish_scripts/util.sh

```diff
@@ -19,110 +19,446 @@
-    done <<< "${commits}"
-    
-    # track the k8s.io/kubernetes commit SHA so we can always determine which level of kube this repo matches
-    # track the filtered branch commit SHA so that we can determine which commits need to be picked
-    echo ${newKubeSHA} > kubernetes-sha
-    if git diff --exit-code &>/dev/null; then
-        echo "SHAs haven't changed!"
-        return
+        local k_base_merge=$(git-find-merge ${k_base_commit} upstream-kube/${src_branch})
+        if [ -z "${k_base_merge}" ]; then
+            echo "Didn't find merge commit of k8s.io/kubernetes commit ${k_base_commit}. Odd."
+            return 1
+        fi
+        git branch -f filtered-branch-base ${k_base_merge} >/dev/null
+        extra_filter_branches="${extra_filter_branches} filtered-branch-base"
+    fi
+
+    # this command rewrites git history to *only* include $subdirectory
+    echo "Rewriting upstream branch ${src_branch} to only include commits for ${subdirectory}."
+    git filter-branch -f --msg-filter 'awk 1 && echo && echo "Kubernetes-commit: ${GIT_COMMIT}"' --subdirectory-filter "${subdirectory}" --tag-name-filter 'sed "s/^/filtered-branch-/"' -- filtered-branch ${extra_filter_branches}
+
+    local f_base_commit=""
+    local f_base_merge_commit=""
+    local f_base_merge_parent=""
+    if [ "${new_repo}" = true ] && [ "${src_branch}" != master ]; then
+        # for a new repo that is not master: map filtered-branch-base to our repo as ${dst_branch_point_commit}
+
+        local k_branch_point_commit=$(kube-commit filtered-branch-base) # this will probably different than the k_branch_point_commit above because filtered drops commits and maps to ancestors if necessary
+        local dst_branch_point_commit=$(branch-commit ${k_branch_point_commit} master)
+        if [ -z "${dst_branch_point_commit}" ]; then
+            # fall-back to the old "sync(k8s.io/kubernetes) <sha1>" commits
+            dst_branch_point_commit=$(git log --oneline --grep "sync(k8s.io/kubernetes) *${k_branch_point_commit}" --format='%H' master)
+            if [ -z "${dst_branch_point_commit}" ]; then
+                echo "Couldn't find a corresponding branch point commit for ${k_branch_point_commit} as ascendent of origin/master."
+                return 1
+            fi
+        fi
+
+        git branch -f ${dst_branch} ${dst_branch_point_commit} >/dev/null
+
+        # there might be merges from master into ${dst_branch}. We pick the last one as the base because everything after is linear again.
+        local filtered_branch_base_merge_commit=""
+        local parent1=""
+        local parent2=""
+        local filtered_branch_base_commit=$(git rev-parse filtered-branch-base)
+        read filtered_branch_base_merge_commit parent1 parent2 <<<$(git log --first-parent --format='%h %P' filtered-branch | grep ${filtered_branch_base_commit})
```

Comment:
the `git log ... | grep` gives the first with `${filtered_branch_base_commit}` as parent. But that's not the first merge on `filtered-branch`. You are right, the following cherry-picks will break if they hit a merge with master.

In general, I think we need faithful support for these master merges. It could be that our branch has been created some days ago (like release-1.8) and the we hit a master merge. I think this will make us fail, at least if the master merge is not a ff merge.

Reply:
this code is gone with faithful master merge support